src/jrd/recsrc/TableValueFunctionScan.cpp

```diff
@@ -0,0 +1,277 @@
+		rpb->rpb_number.setValid(false);
+		return false;
+	}
+	return true;
+}
+
+bool TableValueFunctionScan::refetchRecord(thread_db* /*tdbb*/) const
+{
+	return true;
+}
+
+WriteLockResult TableValueFunctionScan::lockRecord(thread_db* /*tdbb*/) const
+{
+	status_exception::raise(Arg::Gds(isc_record_lock_not_supp));
+}
+
+void TableValueFunctionScan::getLegacyPlan(thread_db* tdbb, Firebird::string& plan,
+										   unsigned level) const
+{
+	if (!level)
+		plan += "(";
+
+	plan += printName(tdbb, m_alias, false) + " NATURAL";
+
+	if (!level)
+		plan += ")";
+}
+
+void TableValueFunctionScan::assignParameter(thread_db* tdbb, dsc* fromDesc, const dsc* toDesc,
+											 SSHORT toId, Record* record) const
+{
+	dsc toDescValue = *toDesc;
+	toDescValue.dsc_address = record->getData() + (IPTR)toDesc->dsc_address;
+
+	if (fromDesc->isNull())
+	{
+		record->setNull(toId);
+		return;
+	}
+	else
+		record->clearNull(toId);
+
+	if (!DSC_EQUIV(fromDesc, &toDescValue, false))
+	{
+		MOV_move(tdbb, fromDesc, &toDescValue);
+		return;
+	}
+
+	memcpy(toDescValue.dsc_address, fromDesc->dsc_address, fromDesc->dsc_length);
+}
+
+UnlistFunctionScan::UnlistFunctionScan(CompilerScratch* csb, StreamType stream, const string& alias,
+									   ValueListNode* list)
+	: TableValueFunctionScan(csb, stream, alias), m_inputList(list)
+{
+}
+
+void UnlistFunctionScan::internalOpen(thread_db* tdbb) const
+{
+	const auto request = tdbb->getRequest();
+	const auto impure = request->getImpure<Impure>(m_impure);
+
+	impure->irsb_flags = irsb_open;
+
+	const auto rpb = &request->req_rpb[m_stream];
+
+	MemoryPool& pool = *tdbb->getDefaultPool();
+	impure->m_recordBuffer = FB_NEW_POOL(pool) RecordBuffer(pool, m_format);
+
+	Record* const record = VIO_record(tdbb, rpb, m_format, &pool);
+
+	rpb->rpb_number.setValue(BOF_NUMBER);
+
+	fb_assert(m_inputList->items.getCount() >= UNLIST_INDEX_LAST);
+
+	auto valueItem = m_inputList->items[UNLIST_INDEX_STRING];
+	const auto valueDesc = EVL_expr(tdbb, request, valueItem);
+	if (valueDesc == nullptr)
+	{
+		rpb->rpb_number.setValid(true);
+		return;
+	}
+
+	auto separatorItem = m_inputList->items[UNLIST_INDEX_SEPARATOR];
+	const auto separatorDesc = EVL_expr(tdbb, request, separatorItem);
+	if (separatorDesc == nullptr)
+	{
+		rpb->rpb_number.setValid(false);
+		return;
+	}
+
+	auto toDesc = m_format->fmt_desc.begin();
+	fb_assert(toDesc);
+
+	const auto textType = toDesc->getTextType();
+
+	auto setStringToRecord = [&] (string str, USHORT length = 0)
+	{
+		if (length == 0)
+			length = str.length();
+		dsc fromDesc;
+		fromDesc.makeText(length, textType, (UCHAR*)str.c_str());
+		assignParameter(tdbb, &fromDesc, toDesc, 0, record);
+		impure->m_recordBuffer->store(record);
+	};
+
+	string separatorStr(MOV_make_string2(tdbb, separatorDesc, textType, true));
+
+	if (separatorStr.isEmpty())
+	{
+		string valueStr(MOV_make_string2(tdbb, valueDesc, textType, false));
+		setStringToRecord(valueStr);
+		return;
+	}
+
+	if (valueDesc->isBlob())
```

Comment:
Others Firebird functions compare strings using collation rules.
Here this is disregard. Should it be?

Reply:
Good question to be discussed. Theoretically, one may ask to split the text into chunks by e.g. both 'A' and 'a' used as a case-insensitive separator 'a'. But honestly, I cannot imagine that being used in practice. And collation-aware compares would add some extra overhead. So maybe it's worth comparing byte-wise (converting both value and separator to the common data type, if necessary). Or maybe we should consider two code branches depending on whether byte-wise compare is possible or not.